File: src/csharp/puzzleImpl/Day01.cs
using static csharpimpl.Helper;

namespace csharpimpl.puzzleImpl;

public class Day01 : IPuzzle
{
    private readonly string[] _input;
    public Day01(string filename)
    {
        _input = File.ReadAllLines(filename);
    }
    
    public string? FirstPuzzle()
    {
        var numbers = _input
            .Select(line => line.Trim())
            .Select(int.Parse).ToList();

        var countIncreases = 0;
        for (int i = 1; i < numbers.Count; i++) {
            if(numbers[i] > numbers[i - 1]) {
                countIncreases++;
            } 
        }

        DebugMsg($"There was {countIncreases} out of total {numbers.Count} numbers.");
        return countIncreases.ToString();
    }

    public string? SecondPuzzle()
    {        
        var numbers = _input
            .Select(line => line.Trim())
            .Select(int.Parse).ToList();

        var zipped = numbers.Zip(numbers.Skip(1), (a, b) => new { a, b })
            .Zip(numbers.Skip(2), (ab, c) =>  (ab.a, ab.b, c));

        var summed = zipped.Select(tuple => tuple.a + tuple.b + tuple.c).ToList();

        var countIncreases = 0;
        for (int i = 1; i < summed.Count; i++) {
            if(summed[i] > summed[i - 1]) {
                countIncreases++;
            } 
        }

        DebugMsg($"There was {countIncreases} out of total {summed.Count} tuples.");
        return countIncreases.ToString();
    }
}

File: src/kotlin/app/adventofcode/impl/Day01.kt
package adventofcode.impl

import adventofcode.Puzzle
import adventofcode.Helper

class Day01(filename: String) : Puzzle(filename) {

    override fun solveFirst(): String? {
        val numbers = input
            .map { it.trim() }
            .map { it.toInt() }

        val increases = countIncreases(numbers)
        Helper.debugMsg("There was $increases out of total ${numbers.size} numbers.");
        return "$increases"
    }

    override fun solveSecond(): String? {

        val numbers = input
            .map { it.trim() }
            .map { it.toInt() }

        val zipped = numbers.windowed(3)
        val summed = zipped.map { it.sum() }

        val increases = countIncreases(summed)
        Helper.debugMsg("There was $increases out of total ${summed.size} tuples.");
        return "$increases"
    }

    private fun <T : Comparable<T>> countIncreases(collection: List<T>): Int = (1 until collection.size).sumOf { (collection[it] > collection[it - 1]).toInt() }

    private fun Boolean.toInt() = if (this) 1 else 0
}
